extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -259,52 +259,54 @@ export class AmpA4A extends AMP.BaseElement {
     // promise chain due to cancel from unlayout, the promise will be rejected.
     this.promiseId_++;
     const promiseId = this.promiseId_;
+    // Shorthand for: reject promise if current chain is out of date.
+    const checkStillCurrent = promiseId => {
+      if (promiseId != this.promiseId_) {
+        throw cancellation();
+      }
+    };
     this.adPromise_ = viewerFor(this.win).whenFirstVisible()
+      // The following block returns the ad URL, if one is available.
       .then(() => {
-        if (promiseId != this.promiseId_) {
-          return Promise.reject(cancellation());
-        }
+        checkStillCurrent(promiseId);
         return this.getAdUrl();
       })
+      // This block returns the (possibly empty) response to the XHR request.
       .then(adUrl => {
-        if (promiseId != this.promiseId_) {
-          return Promise.reject(cancellation());
-        }
+        checkStillCurrent(promiseId);
         this.adUrl_ = adUrl;
         return this.sendXhrRequest_(adUrl);
       })
+      // The following block returns either a valid ad creative or null, if
+      // no valid creative is available.
       .then(fetchResponse => {
-        if (promiseId != this.promiseId_) {
-          return Promise.reject(cancellation());
+        checkStillCurrent(promiseId);
+        if (!fetchResponse || !fetchResponse.arrayBuffer) {
+          return Promise.resolve(null);
```

Comment:
Just return, no need to wrap it in a Promise (we're already inside a `#then` block).